File: src/main/java/org/quantumbadger/redreader/image/AlbumInfo.java
/*******************************************************************************
 * This file is part of RedReader.
 *
 * RedReader is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * RedReader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with RedReader.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package org.quantumbadger.redreader.image;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import org.apache.commons.text.StringEscapeUtils;
import org.quantumbadger.redreader.jsonwrap.JsonArray;
import org.quantumbadger.redreader.jsonwrap.JsonObject;
import org.quantumbadger.redreader.jsonwrap.JsonValue;

import java.util.ArrayList;

public class AlbumInfo {

	@NonNull public final String url;
	@Nullable public final String title;
	@Nullable public final String description;

	@NonNull public final ArrayList<ImageInfo> images;

	public AlbumInfo(
			@NonNull final String url,
			@Nullable final String title,
			@Nullable final String description,
			@NonNull final ArrayList<ImageInfo> images) {
		this.url = url;
		this.title = title;
		this.description = description;
		this.images = new ArrayList<>(images);
	}

	public static AlbumInfo parseImgur(
			final String url,
			final JsonObject object) {

		String title = object.getString("title");
		String description = object.getString("description");

		if(title != null) {
			title = StringEscapeUtils.unescapeHtml4(title);
		}

		if(description != null) {
			description = StringEscapeUtils.unescapeHtml4(description);
		}

		final JsonArray imagesJson = object.getArray("images");
		final ArrayList<ImageInfo> images = new ArrayList<>();

		for(final JsonValue imageJson : imagesJson) {
			images.add(ImageInfo.parseImgur(imageJson.asObject()));
		}

		return new AlbumInfo(url, title, description, images);
	}

	public static AlbumInfo parseImgurV3(
			final String url,
			final JsonObject object) {

		String title = object.getString("title");
		String description = object.getString("description");

		if(title != null) {
			title = StringEscapeUtils.unescapeHtml4(title);
		}

		if(description != null) {
			description = StringEscapeUtils.unescapeHtml4(description);
		}

		final JsonArray imagesJson = object.getArray("images");
		final ArrayList<ImageInfo> images = new ArrayList<>();

		for(final JsonValue imageJson : imagesJson) {
			images.add(ImageInfo.parseImgurV3(imageJson.asObject()));
		}

		return new AlbumInfo(url, title, description, images);
	}

	@NonNull
	private static ImageInfo.MediaType stringToMediaType(@Nullable final String mediaTypeString) {

		if(mediaTypeString == null) {
			return ImageInfo.MediaType.IMAGE;

		} else {
			switch(mediaTypeString) {

				case "AnimatedImage":
					return ImageInfo.MediaType.GIF;

				case "Video":
					// This string doesn't seem to exist yet, but it might do in future
					return ImageInfo.MediaType.VIDEO;

				case "Image":
				default:
					return ImageInfo.MediaType.IMAGE;
			}
		}
	}

	@Nullable
	private static String getThumbnail(final JsonArray images) {

		final int minThumbSize = 200;

		Integer bestSizeMinAxis = null;
		String bestUrl = null;

		for(final JsonValue value : images) {

			final JsonObject image = value.asObject();

			final int x = (int)(long)image.getLong("x");
			final int y = (int)(long)image.getLong("y");

			final int minAxis = Math.min(x, y);

			if(bestSizeMinAxis == null
					|| (bestSizeMinAxis < minThumbSize && minAxis > bestSizeMinAxis)
					|| (minAxis >= minThumbSize && minAxis < bestSizeMinAxis)) {

				bestSizeMinAxis = minAxis;
				bestUrl = image.getString("u");
			}
		}

		return StringEscapeUtils.unescapeHtml4(bestUrl);
	}

	@Nullable
	public static AlbumInfo parseRedditGallery(
			final String url,
			final JsonObject object) {

		final JsonObject mediaMetadataList = object.getObject("media_metadata");
		final JsonObject galleryData = object.getObject("gallery_data");

		if(mediaMetadataList == null || galleryData == null) {
			return null;
		}

		final JsonArray galleryItems = galleryData.getArray("items");

		final ArrayList<ImageInfo> images = new ArrayList<>();

		for(final JsonValue itemValue : galleryItems) {

			final JsonObject item = itemValue.asObject();

			final String mediaId = StringEscapeUtils.unescapeHtml4(item.getString("media_id"));

			@Nullable final String caption
					= StringEscapeUtils.unescapeHtml4(item.getString("caption"));

			// TODO show this in the UI
			@SuppressWarnings("unused")
			@Nullable final String outboundUrl
					= StringEscapeUtils.unescapeHtml4(item.getString("outbound_url"));

			final JsonObject mediaMetadataEntry = mediaMetadataList.getObject(mediaId);

			@Nullable final String mimetype
					= StringEscapeUtils.unescapeHtml4(mediaMetadataEntry.getString("m"));

			final JsonObject standardImage = mediaMetadataEntry.getObject("s");

			final ImageInfo.MediaType mediaType
					= stringToMediaType(mediaMetadataEntry.getString("e"));

			String urlEscaped = standardImage.getString("u");

			if(urlEscaped == null) {
				urlEscaped = standardImage.getString("mp4");

				if(urlEscaped == null) {
					urlEscaped = standardImage.getString("gif");
				}
			}

			images.add(new ImageInfo(
					StringEscapeUtils.unescapeHtml4(urlEscaped),
					getThumbnail(mediaMetadataEntry.getArray("p")),
					caption,
					null,
					mimetype,
					mediaType != ImageInfo.MediaType.IMAGE,
					standardImage.getLong("x"),
					standardImage.getLong("y"),
					null,
					mediaType,
					ImageInfo.HasAudio.NO_AUDIO,
					null,
					null,
					null));
		}

		final String title = StringEscapeUtils.unescapeHtml4(object.getString("title"));

		return new AlbumInfo(url, title, null, images);
	}
}

File: src/main/java/org/quantumbadger/redreader/image/AlbumInfo.kt
/*******************************************************************************
 * This file is part of RedReader.
 *
 * RedReader is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * RedReader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with RedReader.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package org.quantumbadger.redreader.image

import androidx.compose.runtime.Immutable
import org.apache.commons.text.StringEscapeUtils
import org.quantumbadger.redreader.common.UriString
import org.quantumbadger.redreader.jsonwrap.JsonObject
import org.quantumbadger.redreader.reddit.kthings.ImageMetadata
import org.quantumbadger.redreader.reddit.kthings.MaybeParseError
import org.quantumbadger.redreader.reddit.kthings.RedditPost
import kotlin.math.min

@Immutable
class AlbumInfo(
	@JvmField val url: UriString,
	@JvmField val title: String?,
	val description: String?,
	images: List<ImageInfo>
) {
	@JvmField
	val images = ArrayList(images)

	companion object {
		@JvmStatic
		fun parseImgur(
			url: UriString,
			obj: JsonObject
		): AlbumInfo {
			var title = obj.getString("title")
			var description = obj.getString("description")

			if (title != null) {
				title = StringEscapeUtils.unescapeHtml4(title)
			}

			if (description != null) {
				description = StringEscapeUtils.unescapeHtml4(description)
			}

			val imagesJson = obj.getArray("images")
			val images = ArrayList<ImageInfo>()

			for (imageJson in imagesJson!!) {
				images.add(ImageInfo.parseImgur(imageJson.asObject()))
			}

			return AlbumInfo(url, title, description, images)
		}

		@JvmStatic
		fun parseImgurV3(
			url: UriString,
			obj: JsonObject
		): AlbumInfo {
			var title = obj.getString("title")
			var description = obj.getString("description")

			if (title != null) {
				title = StringEscapeUtils.unescapeHtml4(title)
			}

			if (description != null) {
				description = StringEscapeUtils.unescapeHtml4(description)
			}

			val imagesJson = obj.getArray("images")
			val images = ArrayList<ImageInfo>()

			for (imageJson in imagesJson!!) {
				images.add(ImageInfo.parseImgurV3(imageJson.asObject()))
			}

			return AlbumInfo(url, title, description, images)
		}

		private fun stringToMediaType(mediaTypeString: String?): ImageInfo.MediaType {
			return if (mediaTypeString == null) {
				ImageInfo.MediaType.IMAGE
			} else {
				when (mediaTypeString) {
					"AnimatedImage" -> ImageInfo.MediaType.GIF
					// This string doesn't seem to exist yet, but it might do in future
					"Video" -> ImageInfo.MediaType.VIDEO
					"Image" -> ImageInfo.MediaType.IMAGE
					else -> ImageInfo.MediaType.IMAGE
				}
			}
		}

		private fun getPreview(
			minSizePx: Int,
			images: List<ImageMetadata>
		): ImageUrlInfo? {

			var bestSizeMinAxis: Long? = null
			var bestUrl: String? = null
			var bestSize: ImageSize? = null

			for (image in images) {

				if (image.u == null) {
					continue
				}

				val x = image.x
				val y = image.y

				val minAxis = min(x, y)

				if (bestSizeMinAxis == null || (bestSizeMinAxis < minSizePx && minAxis > bestSizeMinAxis)
					|| (minAxis >= minSizePx && minAxis < bestSizeMinAxis)
				) {
					bestSizeMinAxis = minAxis
					bestUrl = image.u.decoded
					bestSize = ImageSize.from(image.x, image.y)
				}
			}

			return bestUrl?.let {
				ImageUrlInfo(
					UriString(bestUrl),
					size = bestSize
				)
			}
		}

		fun parseRedditGallery(post: RedditPost): AlbumInfo? {

			val galleryItems = post.gallery_data?.items ?: return null

			val images = galleryItems.mapNotNull { (it as? MaybeParseError.Ok)?.value }
				.mapNotNull { item ->

					val mediaMetadataEntry =
						(post.media_metadata?.get(item.media_id) as? MaybeParseError.Ok)?.value
							?: return@mapNotNull null

					val standardImage = mediaMetadataEntry.s

					val mediaType = stringToMediaType(mediaMetadataEntry.e)

					val urlEscaped = standardImage.u ?: standardImage.mp4 ?: standardImage.gif

					val original = urlEscaped?.let {
						ImageUrlInfo(
							url = UriString(urlEscaped.decoded),
							size = ImageSize.from(standardImage.x, standardImage.y)
						)
					}

					if (original == null) {
						throw RuntimeException("url missing from response")
					}

					var bigSquare: ImageUrlInfo? = null
					var preview: ImageUrlInfo? = null

					mediaMetadataEntry.p?.let { p ->
						val images = p + listOf(standardImage)

						bigSquare = getPreview(minSizePx = 300, images = images)
						preview = getPreview(minSizePx = 1400, images = images)
					}

					ImageInfo(
						original = original,
						preview = preview,
						bigSquare = bigSquare,
						title = item.caption?.decoded,
						outboundUrl = UriString.fromNullable(
							item.outbound_url?.decoded?.trim()?.takeUnless { it.isEmpty() }),
						type = mediaMetadataEntry.m,
						isAnimated = mediaType != ImageInfo.MediaType.IMAGE,
						mediaType = mediaType,
						hasAudio = ImageInfo.HasAudio.MAYBE_AUDIO,
					)
				}.toList()

			val title = post.title?.decoded

			return AlbumInfo(post.findUrl()!!, title, null, images)
		}
	}
}
